File: go/isValidSudoku.go
// https://leetcode.com/problems/valid-sudoku

package leetcode

func isValidSudoku(board [][]byte) bool {
	rows := [9][9]bool{}
	columns := [9][9]bool{}
	squares := [3][3][9]bool{}

	for row := 0; row < 9; row++ {
		for col := 0; col < 9; col++ {
			cell := board[row][col]

			if cell == '.' {
				continue
			}

			digit := int(cell-'0') - 1

			if rows[row][digit] || columns[col][digit] || squares[row/3][col/3][digit] {
				return false
			}

			rows[row][digit] = true
			columns[col][digit] = true
			squares[row/3][col/3][digit] = true
		}
	}

	return true
}

File: python/src/isValidSudoku.py
# https://leetcode.com/problems/valid-sudoku

from collections import defaultdict
from typing import List


class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        rows = defaultdict(set)
        cols = defaultdict(set)
        boxs = defaultdict(set)

        for row in range(9):
            for col in range(9):
                box = (row // 3, col // 3)
                cell = board[row][col]
                if cell == ".":
                    continue

                if cell in rows[row] or cell in cols[col] or cell in boxs[box]:
                    return False

                rows[row].add(cell)
                cols[col].add(cell)
                boxs[box].add(cell)

        return True
